wideq/washer.py

```diff
@@ -0,0 +1,123 @@
+import enum
+from typing import Optional
+
+from .client import Device
+
+
+class WasherState(enum.Enum):
+    """The state of the washer device."""
+
+    ADD_DRAIN = '@WM_STATE_ADD_DRAIN_W'
+    DETECTING = '@WM_STATE_DETECTING_W'
+    DRYING = '@WM_STATE_DRYING_W'
+    END = '@WM_STATE_END_W'
+    ERROR_AUTO_OFF = '@WM_STATE_ERROR_AUTO_OFF_W'
+    FRESH_CARE = '@WM_STATE_FRESHCARE_W'
+    INITIAL = '@WM_STATE_INITIAL_W'
+    OFF = '@WM_STATE_POWER_OFF_W'
+    PAUSE = '@WM_STATE_PAUSE_W'
+    PRE_WASH = '@WM_STATE_PREWASH_W'
+    RESERVE = '@WM_STATE_RESERVE_W'
+    RINSING = '@WM_STATE_RINSING_W'
+    RINSE_HOLD = '@WM_STATE_RINSE_HOLD_W'
+    RUNNING = '@WM_STATE_RUNNING_W'
+    SMART_DIAGNOSIS = '@WM_STATE_SMART_DIAG_W'
+    SMART_DIAGNOSIS_DATA = '@WM_STATE_SMART_DIAGDATA_W'
+    SPINNING = '@WM_STATE_SPINNING_W'
+    TUBCLEAN_COUNT_ALARM = '@WM_STATE_TUBCLEAN_COUNT_ALRAM_W'
+
+
+class WasherDevice(Device):
+    """A higher-level interface for a washer."""
+
+    def poll(self) -> Optional['WasherStatus']:
+        """Poll the device's current state.
+
+        Monitoring must be started first with `monitor_start`.
+
+        :returns: Either a `WasherStatus` instance or `None` if the status is
+            not yet available.
+        """
+        # Abort if monitoring has not started yet.
+        if not hasattr(self, 'mon'):
+            return None
+
+        res = self.mon.poll_json()
+        if res:
+            return WasherStatus(self, res)
+        else:
+            return None
+
+
+class WasherStatus(object):
+    """Higher-level information about a washer's current status.
+
+    :param washer: The WasherDevice instance.
+    :param data: JSON data from the API.
+    """
+
+    def __init__(self, washer: WasherDevice, data: dict):
+        self.washer = washer
+        self.data = data
+
+    def _lookup_enum(self, attr: str) -> str:
```

Comment:
Because this looks like the same method as in `wideq.dryer`, maybe we should find a way to centralize utilities like this?